### PR TITLE
Add backup/restore tutorial

### LINKED ISSUE #746: Data backup tutorial
We should create a tutorial outlining the techniques necessary to take regular backups of data volume containers and store them somewhere less volatile than Carina. We could use a technique similar to the [cron container in NGINX/Let's Encrypt](https://getcarina.com/docs/tutorials/nginx-with-lets-encrypt/#reissue-certificates-with-a-cron-container) to connect directly to the data volume containers, `docker cp` the contents, tarball them, then use `rack` or vanilla cURL to upload them to Cloud Files.

* [x] Be tiny (based on Alpine)
* [x] Support backing up multiple volumes
* [x] Support saving backup directly to client instead of uploading to CF
* [x] Write tutorial content for doing backup/restore
* [ ] Write tutorial content for scheduling container runs and other tasks.
* [ ] Replace https://getcarina.com/docs/tutorials/quassel-on-carina/#automatically-back-up-your-data-and-restore-it-as-necessary section with links to aforementioned tutorials.

### PR DESCRIPTION
**Pull request by ktbartholomew**
Closes #746.

### CODE REVIEW

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+  - intermediate
+---
+
+This tutorial explains how to back up and restore data from your containers, so you can quickly restore your application data in the event of a system failure.
+
+### Prerequisites
+
+Before you begin, you need to [create and connect to a Carina cluster]({{ site.baseurl }}/docs/tutorials/create-connect-cluster/).
+
+Optionally, if you would like to store your backups in [Rackspace Cloud Files](https://www.rackspace.com/cloud/files), you need a traditional Rackspace Cloud account (one _not_ created from the Carina website).
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: For consistency with other tutorials, the text for these Prereqs should be 

```
* [Create and connect to a cluster](/docs/tutorials/create-connect-cluster/)
* _(Optional)_ A Rackspace cloud account that you can use to access the [Cloud Control Panel](https://mycloud.rackspace.com/).
 * If you don't have a Rackspace cloud account, you need to [sign up for one](https://www.rackspace.com/cloud).
```

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+### Create a MySQL instance
+
+In order to effectively test a backup process, you’ll need something worth backing up. In this section, you’ll create a MySQL instance with a separate [data volume container](/docs/tutorials/data-volume-containers/) for its data storage.
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: We avoid linking in the body of a tutorial unless we _need_ the reader to do something or get some critical info. Keep the text but add the link to the Resources section.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+1. Create an [overlay network]({{ site.baseurl }}/docs/tutorials/overlay-networks/) named `mysql` to allow multiple containers to communicate with one another.
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: We avoid linking in the body of a tutorial unless we _need_ the reader to do something or get some critical info. Keep the text but add the link to the Resources section.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+      --detach \
+      --env MYSQL_ROOT_PASSWORD=secret \
+      --env MYSQL_DATABASE=test \
+      --name mysql-server \
+      --net mysql \
+      --volumes-from mysql-data \
+      mysql:latest
+    ```
+
+1. Populate the database with some empty tables for testing purposes (we’re borrowing the `carinamarina/guestbook-mysql` Docker image used in the tutorial [Use MySQL on Carina]({{ site.baseurl }}/docs/tutorials/data-stores-mysql/)).
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: We avoid linking in the body of a tutorial unless we _need_ the reader to do something or get some critical info. Keep the text but add the link to the Resources section.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+      --env MYSQL_PASSWORD=secret \
+      --env MYSQL_DATABASE=test \
+      --net mysql \
+      carinamarina/guestbook-mysql \
+      python app.py create_tables
+    ```
+
+### Back up the database
+
+Now that you’ve created a MySQL instance with some data, you can use our [`carinamarina/backup`](https://hub.docker.com/r/carinamarina/backup/) Docker image to store the data in a safe place.
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: We avoid linking in the body of a tutorial unless we _need_ the reader to do something or get some critical info. Keep the text but add the link to the Resources section.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+      carinamarina/backup \
+      backup \
+      --source /backups/ \
+      --container <name-of-cloud-files-container> \
+      --zip
+    Bundling archive...
+    Uploading archive to Cloud Files...
+    Finished! Uploaded object [2016/03/11/21-30-backups.tar.gz] to container [carina-backup] in now
+    Done.
+    ```
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]:
```bash
$ docker run \
>   --rm \
>   --env RS_USERNAME=$RS_USERNAME \
>   --env RS_API_KEY=$RS_API_KEY \
>   --env RS_REGION_NAME=IAD \
>   --volumes-from mysql-data \
>   carinamarina/backup \
>   backup \
>   --source /backups/ \
>   --container my-sql-backups \
>   --zip
Bundling archive...
Uploading archive to Cloud Files...
error	I couldn't find container [my-sql-backups]
```

The script in the image should create the CF-container if it doesn't exist.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    ```bash
+    $ docker run \
+      --rm \
+      --net mysql \
+      --volumes-from mysql-data \
+      mysql \
+      bash -c "mysqldump -psecret -h mysql-server --databases test > /backups/test.sql"
+    ```
+
+1. Back up the database dump to your local filesystem.
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: Let's create a separate section for this. I can see me linking to this specific thing very often.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+      carinamarina/backup \
+      backup \
+      --source /backups/ \
+      --stdout \
+      --zip > my-local-backup.tar.gz
+    ```
+
+    This adds all the contents of `/backups/` from your data volume container to a compressed <a href="https://en.wikipedia.org/wiki/Tar_(computing)"><code>tar</code></a> archive and pipe it to a file on your local filesystem. Whatever you do with the backup file after this is up to you.
+
+1. _(Optional)_ Back up the database dump to Rackspace Cloud Files. If you have a paid Rackspace account in addition to your Carina account, you can store your backups in a Cloud Files container.
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: Let's create a separate section for this. I can see me linking to this specific thing very often.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+    ```
+
+1. Create a data volume container named `mysql-data`. The MySQL image stores its table data in `/var/lib/mysql`, so include `--volume /var/lib/mysql` to present that path as a volume that other containers can mount. Include `--volume /backups` to create a persistent folder for your database backups.
+
+    ```bash
+    $ docker create \
+      --name mysql-data \
+      --volume /var/lib/mysql \
+      --volume /backups \
+      mysql:latest
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: Some places you use `mysql` some `mysql:latest` should be consistent throughout the tutorial.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    ```bash
+    $ docker run \
+      --rm \
+      --env RS_USERNAME=<your-rackspace-username> \
+      --env RS_API_KEY=<your-rackspace-api-key> \
+      --env RS_REGION_NAME=IAD \
+      --volumes-from mysql-data \
+      carinamarina/backup \
+      restore \
+      --container carina-backup \
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: `--container <name-of-cloud-files-container>`

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+      --object 2016/03/11/21-30-backups.tar.gz \
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: Describe how this path comes from the backup command above and is shown in its output. (`Finished! Uploaded object [2016/03/11/21-30-backups.tar.gz] to container`)

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+permalink: docs/tutorials/backup-restore-data/
+description: Back up and restore the contents of data volume containers
+docker-versions:
+  - 1.10.1
+topics:
+  - docker
+  - intermediate
+---
+
+This tutorial explains how to back up and restore data from your containers, so you can quickly restore your application data in the event of a system failure.
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: Maybe add

"This tutorials describes how to backup a MySQL database in detail but the concepts are applicable to backing up any files in a container."

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+      --volumes-from mysql-data \
+      carinamarina/backup \
+      restore \
+      --container carina-backup \
+      --object 2016/03/11/21-30-backups.tar.gz \
+      --destination /backups/
+      --zip
+    Reading and unzipping archive...
+    Done.
+    ```
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]:
```
$ docker run \
>   --rm \
>   --env RS_USERNAME=$RS_USERNAME \
>   --env RS_API_KEY=$RS_API_KEY \
>   --env RS_REGION_NAME=IAD \
>   --volumes-from mysql-data \
>   carinamarina/backup \
>   restore \
>   --container my-sql-backups \
>   --object 2016/03/11/21-11-backups.tar.gz \
>   --destination /backups/ \
>   --zip
Reading and unzipping archive...
error	Resource not found
tar: invalid magic
tar: short read
```

I got the CF-container path wrong on my first try and got that error message. Please add that error message to the Troubleshooting section and what to do about it.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+      --rm \
+      --env RS_USERNAME=<your-rackspace-username> \
+      --env RS_API_KEY=<your-rackspace-api-key> \
+      --env RS_REGION_NAME=IAD \
+      --volumes-from mysql-data \
+      carinamarina/backup \
+      restore \
+      --container carina-backup \
+      --object 2016/03/11/21-30-backups.tar.gz \
+      --destination /backups/
```
Comment (everett-toews) [inline _tutorials/2016-03-14-backup-restore-data.md]: Missing trailing `\`

**Comment by everett-toews** [comment]
screen shot 2016-03-11 at 5 12 06 pm

This is really awesome!

You should include a nice Cloud Files screen shot just before the Prereqs section to add some colour.

File: _tutorials/2016-03-14-backup-restore-data.md (hunk 6 above)
Comment (ktbartholomew) [inline _tutorials/2016-03-14-backup-restore-data.md]: https://github.com/getcarina/examples/issues/10

File: _tutorials/2016-03-14-backup-restore-data.md (hunk 13 above)
Comment (ktbartholomew) [inline _tutorials/2016-03-14-backup-restore-data.md]: Fixed w/ https://github.com/getcarina/examples/pull/11

File: _tutorials/2016-03-14-backup-restore-data.md (hunk 9 above)
Comment (ktbartholomew) [inline _tutorials/2016-03-14-backup-restore-data.md]: Changed to `mysql:latest` throughout

**Comment by everett-toews** [comment]
@kmsholcomb Can you please review this as soon as possible?

We need this for some customer communications. Thanks.

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+      --rm \
+      --volumes-from mysql-data \
+      carinamarina/backup \
+      backup \
+      --source /backups/ \
+      --stdout \
+      --zip > my-local-backup.tar.gz
+    ```
+
+    This adds all the contents of `/backups/` from your data volume container to a compressed tar archive and pipe it to a file on your local filesystem. Whatever you do with the backup file after this is up to you.
```
Comment (catlook) [inline _tutorials/2016-03-14-backup-restore-data.md]: ... archive and pipes it to ...

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+      restore \
+      --destination /backups/ \
+      --stdin \
+      --zip \
+      < my-local-backup.tar.gz
+    Reading and unzipping archive...
+    Done.
+    ```
+
+1. _(Optional)_ Download the backup archive from Cloud Files and unpack it to the `/backups/` volume.
```
Comment (catlook) [inline _tutorials/2016-03-14-backup-restore-data.md]: ... Cloud Files, and unpack  ...

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    ```bash
+    $ docker run \
+      --rm \
+      --net mysql \
+      --volumes-from mysql-data \
+      mysql:latest \
+      bash -c "mysql -p<your-mysql-password> -h mysql-server < /backups/test.sql"
+    ```
+
+That's all! You’ve successfully created a MySQL server, backed up the contents of a database to a compressed archive, then restored data using your backup.
```
Comment (catlook) [inline _tutorials/2016-03-14-backup-restore-data.md]: ... archive, and then ...

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+
+### Troubleshooting
+
+See [Troubleshooting common problems]({{site.baseurl}}/docs/troubleshooting/common-problems/).
+
+For additional assistance, ask the [community](https://community.getcarina.com/) for help or join us in IRC at [#carina on Freenode](http://webchat.freenode.net/?channels=carina).
```
Comment (catlook) [inline _tutorials/2016-03-14-backup-restore-data.md]: ... help, or join us ...

File: _tutorials/2016-03-14-backup-restore-data.md [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+### Resources
+
+* [`carinamarina/backup` README](https://hub.docker.com/r/carinamarina/backup/)
+* [data volume containers](/docs/tutorials/data-volume-containers/)
+* [overlay networks]({{ site.baseurl }}/docs/tutorials/overlay-networks/)
+* <a href="https://en.wikipedia.org/wiki/Tar_(computing)"><code>Tar</code></a>
+
+### Next step
+
+* [Schedule regular backups using cron]({{ site.baseurl }}/docs/tutorials/schedule-tasks-cron/)
```
Comment (catlook) [inline _tutorials/2016-03-14-backup-restore-data.md]: [Ed note: done!  ]

**Comment by rackernexus** [comment]
If you want to double-check your formatting, check [the preview](https://staging.developer.rackspace.com/getcarina.com/build-04d08376fe/).

**Comment by kmsholcomb** [comment]
Hi @ktbartholomew. I know I'm a little late on this, but I wanted to suggest a few edits. See the attached Word doc (or PDF copy of it). Thanks! 

[Back up and restore container data_editsKH.docx](https://github.com/getcarina/getcarina.com/files/178615/Back.up.and.restore.container.data_editsKH.docx)
[Back up and restore container data_editsKH.pdf](https://github.com/getcarina/getcarina.com/files/178614/Back.up.and.restore.container.data_editsKH.pdf)

**Pull request merged**